src/target/riscv/riscv.c

```diff
@@ -1133,27 +1161,30 @@ static int maybe_add_trigger_t3(struct target *target, bool vs, bool vu,
 	tdata1 = set_field(tdata1, CSR_ICOUNT_U, u);
 	tdata1 = set_field(tdata1, CSR_ICOUNT_COUNT, count);
 
-	unsigned int idx = 0;
-	ret = find_next_free_trigger(target, CSR_TDATA1_TYPE_ICOUNT, false, &idx);
-	if (ret != ERROR_OK)
-		return ret;
-	ret = set_trigger(target, idx, tdata1, 0, 0);
-	if (ret != ERROR_OK)
-		return ret;
-	r->trigger_unique_id[idx] = unique_id;
-	return ERROR_OK;
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+	for (unsigned int idx = 0;
+			find_next_free_trigger(target, CSR_TDATA1_TYPE_ICOUNT,
+					/* chained */ false, &idx) == ERROR_OK;
+			++idx) {
+		ret = sdtrig_set_t1(target, idx, tdata1, /* tdata1 ignore mask */ 0);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = unique_id;
+			return ERROR_OK;
+		}
+		if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			break;
+	}
```

Comment:
The loop can be re-written to avoid the complex expressions inside the `for`. I also tried to add few more comments to make the code easier to grasp for newcomers.

The resulting code is bit longer but hopefully easier to read.

Please let me know if you like it.

```suggestion
	unsigned int idx = 0;
	while (true) {
		int ret = find_next_free_trigger(target, CSR_TDATA1_TYPE_ICOUNT,
					/* chained */ false, &idx);
		if (ret != ERROR_OK) {
			/* No more free triggers, or an error occurred. */
			return ret;
		}
		
		/* A free trigger has been found whose tselect index is now in idx. */
		ret = sdtrig_set_t1(target, idx, tdata1, /* tdata1 ignore mask */ 0);
		if (ret == ERROR_OK) {
			/* Success: the trigger has been set */
			r->trigger_unique_id[idx] = unique_id;
			return ERROR_OK;
		}
		
		if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
			/* An error occurred during setting the trigger */
			return ret;
		
		/* The trigger did not support the options that we required.
		 * Try the next free trigger. */
		 idx++;
	}
```

Reply:
I think it is definitely easier to read, though my personal aesthetic feelings are hurt whenever I see something like `while(true)/while(1)`. I'll try to piggy-back on your suggestion and either adopt it or rework a little bit